src/drivers/mediatek/mt8195/afe-drv.c

```diff
@@ -0,0 +1,414 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Mediatek
+//
+// Author: Bo Pan <bo.pan@mediatek.com>
+//         YC Hung <yc.hung@mediatek.com>
+
+#include <sof/common.h>
+#include <sof/lib/io.h>
+#include <sof/lib/alloc.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <sof/lib/uuid.h>
+#include <errno.h>
+#include <sof/drivers/afe-drv.h>
+#include <sof/drivers/afe-memif.h>
+
+/*#define AFE_DRV_LOG*/
+static struct mtk_base_afe mtk_afe;
+
+#ifndef AFE_DRV_LOG
+#undef printf
+#define printf(format, ...)
+#endif
+
+DECLARE_SOF_UUID("afedrv", afedrv_uuid, 0x4e8f16d1, 0xe935, 0x41f4,
+		 0xb9, 0x9e, 0x42, 0xc5, 0x7e, 0x74, 0x57, 0x84);
+
+DECLARE_TR_CTX(afedrv_tr, SOF_UUID(afedrv_uuid), LOG_LEVEL_INFO);
+
+static inline void afe_reg_read(struct mtk_base_afe *afe, uint32_t reg, uint32_t *value)
+{
+	*value = io_reg_read((uint32_t)((char *)afe->base + reg));
+	tr_dbg(&afedrv_tr, "r_reg:0x%x, value:0x%x\n", reg, *value);
+}
+
+static inline void afe_reg_write(struct mtk_base_afe *afe, uint32_t reg, uint32_t value)
+{
+	io_reg_write((uint32_t)((char *)afe->base + reg), value);
+	tr_dbg(&afedrv_tr, "w_reg:0x%x, value:0x%x\n", reg, value);
+}
+
+static inline void afe_reg_update_bits(struct mtk_base_afe *afe, uint32_t reg, uint32_t mask,
+				       uint32_t value)
+{
+	io_reg_update_bits((uint32_t)((char *)afe->base + reg), mask, value);
+	tr_dbg(&afedrv_tr, "u_reg:0x%x, value:0x%x\n", reg, value);
+}
+
+static int afe_memif_set_channel(struct mtk_base_afe *afe, int id, unsigned int channel)
+{
+	struct mtk_base_afe_memif *memif = &afe->memif[id];
+	unsigned int mono;
+
+	if (memif->data->mono_shift < 0)
+		return 0;
+
+	if (memif->data->ch_num_reg >= 0) {
+		afe_reg_update_bits(afe, memif->data->ch_num_reg,
+				    memif->data->ch_num_maskbit << memif->data->ch_num_shift,
+				    channel << memif->data->ch_num_shift);
+	}
+
+	if (memif->data->quad_ch_mask) {
+		unsigned int quad_ch = channel == 4;
+
+		afe_reg_update_bits(afe, memif->data->quad_ch_reg,
+				    memif->data->quad_ch_mask << memif->data->quad_ch_shift,
+				    quad_ch << memif->data->quad_ch_shift);
+	}
+
+	mono = memif->data->mono_invert ^ channel;
```

Comment:
ok, so, now if `channel` is even then `mono` will be set to 2, 3, 4, or 5. I understand that the mask below is just one low bit, but... Well, fine with me but a bit unclean, it isn't very clear what a `mono` value of 5 means